EmaPajic/assignment3/test_assignment3.py

```diff
@@ -0,0 +1,48 @@
+#!/usr/bin/env python3
+# -*- coding: utf-8 -*-
+"""
+Created on Tue Jun 19 13:44:29 2018
+
+@author: user
+"""
+
+import unittest
+from assignment3 import create_grid,find_words_from_grid
+from dictionary import Dictionary
+
+class TestQ2(unittest.TestCase):
+    
+    def setUp(self):
+        self.empty_dictionary = Dictionary([])
+        self.dictionary1 = Dictionary(['ana','ema','eganam'])
+        self.dictionary2 = Dictionary(['ana','ema','ama','ame'])
+        self.grid = create_grid(3,3,['e','m','a','g','a','n','g','g','g'])   
+        self.word = ""
+        self.foundWords = set()
+        self.visited = {}
+        for x in range(0, 3):
+            for y in range(0, 3):
+                self.visited[(x, y)] = 0
+        
+    def test_empty(self):
+        find_words_from_grid(self.grid,0,0,3,3,self.word,self.empty_dictionary,self.foundWords,self.visited)
+        self.assertEqual(set(),self.foundWords)
+        
+    def test_normal(self):
+        find_words_from_grid(self.grid,0,0,3,3,self.word,self.dictionary1,self.foundWords,self.visited)
```

Comment:
What about turning the function find_words_from_grid() into a closure? This means that it creates all parameters it can from inside the function itself.

E.g.
```
def find_words_from_grid(grid, i, j, n, m, dictionary, word=None, foundWords=None, visited=None):
    if word is None:
        word = ""
    if foundWords is None:
        foundWords = set()
    if visited is None:
        visited = {}
    ...
```

With closures, the person reading the code doesn't lose context by looking at the extra structure initializations outside the function but can focus on the general high-level algorithm.

Reply:
updated :)